### PR TITLE
Ported to cmake build system.

### PR DESCRIPTION
**Pull request by akornatskyy**
# Work In Progress

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+# Paths that are searched when `include(...)` is called.
+list(APPEND CMAKE_MODULE_PATH ${MESOS_ROOT}/cmake)
+list(APPEND CMAKE_MODULE_PATH ${MESOS_ROOT}/3rdparty/cmake)
+list(APPEND CMAKE_MODULE_PATH ${MESOS_ROOT}/src/cmake)
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake)
+
+# Macros.
+include(Versions)
+include(Dependencies)
+#include(cotire)
```
Comment (greggomann) [inline CMakeLists.txt]: Can we get rid of this?

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+add_subdirectory(dockercfg)
+add_subdirectory(journald)
+add_subdirectory(logsink)
+add_subdirectory(metrics)
+
+# FIXME: due to import path in 'overlay/messages.proto'
+# NOTE: according to [1], overlay module is not required for Windows
+# [1] https://github.com/mesosphere/dcos-enterprise/tree/master/packages/mesos/extra/windows-ee.patches
+#add_subdirectory(overlay)
```
Comment (greggomann) [inline CMakeLists.txt]: I think we can get rid of this?

File: cmake/Dependencies.cmake [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+find_library(
+  MESOS_LIB
+  NAMES mesos
+  PATHS ${MESOS_ROOT}/build/src
+  PATH_SUFFIXES Debug Release
+)
+
+SET(Protobuf_DIR ${MESOS_ROOT}/build/3rdparty/protobuf-${PROTOBUF_VERSION}/src/protobuf-${PROTOBUF_VERSION}-build/cmake)
+find_package(Protobuf CONFIG REQUIRED)
```
Comment (greggomann) [inline cmake/Dependencies.cmake]: When I attempt to build locally cmake has some issue finding valid config for protobuf, does this work OK for you guys?

File: overlay/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,56 @@
+
```
Comment (greggomann) [inline overlay/CMakeLists.txt]: If we're not going to include the overlay module in the cmake build for now, perhaps we should just remove this file?

File: CMakeLists.txt (hunk 2 above)
Comment (akornatskyy) [reply CMakeLists.txt]: preserved for linux

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,198 @@
+
+if (NOT WIN32)
+  cmake_minimum_required(VERSION 3.7.0)
+else ()
+  cmake_minimum_required(VERSION 3.8.0)
```
Comment (janisz) [inline CMakeLists.txt]: Can we just stick with 3.8?

Comment (akornatskyy) [reply CMakeLists.txt]: we want to be consistent with mesos build. it uses 3.7.0 for linux.

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+add_subdirectory(journald)
+add_subdirectory(logsink)
+
+# FIXME: due to import path in 'overlay/messages.proto'
+# NOTE: according to [1], overlay module is not required for Windows
+# [1] https://github.com/mesosphere/dcos-enterprise/tree/master/packages/mesos/extra/windows-ee.patches
+#add_subdirectory(overlay)
+
+
+# FIXME: due to limitation of protobuf generation process
```
Comment (asekretenko) [inline CMakeLists.txt]: Could you please check that all FIXMEs (in both opensource and EE PRs) are either resolved now or converted into  TODOs that clearly state what exactly should be fixed and what the desired state is?

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+  add_definitions(
+    -DMODULES_BUILD_DIR="${PROJECT_BINARY_DIR}"
+    -DSOURCE_DIR="${MESOS_ROOT}"
+    -DBUILD_DIR="${MESOS_ROOT}/build"
+    -DPKGMODULEDIR="nowhere"
+    -DLIBDIR="nowhere"
+    -DPKGLIBEXECDIR="nowhere"
+    -DTESTLIBEXECDIR="nowhere"
+    -DSBINDIR="nowhere"
+    -DPKGDATADIR="nowhere"
```
Comment (asekretenko) [inline CMakeLists.txt]: Why do we need to define these at all? 

Is it because something (what exactly?) depends on these definitions existing but doesn't actually use a value?
Or is it just because something will break in a non-transparent way?

In any case, I would ask you to leave a comment explaining the need for these definitions, so that readers don't get lost in the future.

File: metrics/isolator.hpp [hunk 8]
```diff
@@ -170,6 +170,7 @@ class MetricsIsolator : public mesos::slave::Isolator
 } // namespace dcos {
 } // namespace mesosphere {
 
-extern mesos::modules::Module<mesos::slave::Isolator> com_mesosphere_dcos_MetricsIsolatorModule;
+extern "C" mesos::modules::Module<mesos::slave::Isolator>
```
Comment (asekretenko) [inline metrics/isolator.hpp]: Do you have a good explanation why we need "C" linkage here? Is it something related to MSVC, to Windows... or to cmake? 

Can you add a comment or, in case you don't have a good explanation, a TODO so that we figure this out later?

File: overlay/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  "${CURL_INCL}"
+  "${ZOOKEEPER_INCL}"
+  "${MESOS_ROOT}/build/include"
+  "${MESOS_ROOT}/include"
+  # FIXME: due to limitation of protobuf generation process
+  # and include in source
+  ${CMAKE_CURRENT_BINARY_DIR}/..
+  # FIXME: due to the following in agent.hpp
+  # #include <overlay/agent_metrics.hpp>
+  # #include <overlay/messages.hpp>
```
Comment (asekretenko) [inline overlay/CMakeLists.txt]: Can this be fixed in "agent.hpp"?

File: overlay/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  ${CMAKE_CURRENT_SOURCE_DIR}/..
+)
+
+add_library(overlay SHARED
+  agent_metrics.cpp
+  agent.cpp
+  # LNK2019: unresolved external symbol "public: __cdecl mesos:: log::Log::Log
+  # `log` is explicitly excluded for Windows in mesos
+  # master_metrics.cpp
+  # master.cpp
```
Comment (asekretenko) [inline overlay/CMakeLists.txt]: Took me some time out that this comment refers to Mesos replicated log that is not supported on Windows; could you rephrase this comment to make this more clear?

File: cmake/Dependencies.cmake [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+  PATHS ${MESOS_ROOT}/build/src
+  PATH_SUFFIXES Release
+)
+find_library(
+  MESOS_LIB_DEBUG
+  NAMES mesos
+  PATHS ${MESOS_ROOT}/build/src
+  PATH_SUFFIXES Debug
+)
+add_library(mesos STATIC IMPORTED)
```
Comment (asekretenko) [inline cmake/Dependencies.cmake]: Do I understand correctly that this statically links Mesos into each dll? 
If yes, then what is the resulting size of each dll?

File: CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,198 @@
+
+if (NOT WIN32)
+  cmake_minimum_required(VERSION 3.7.0)
+else ()
+  cmake_minimum_required(VERSION 3.8.0)
+endif ()
+
+project(dcos-mesos-modules)
+
+get_filename_component(MESOS_ROOT ../mesos ABSOLUTE)
```
[the remark this answers is not in the document]
Comment (akornatskyy) [reply CMakeLists.txt]: Yes, done.

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+include(Dependencies)
+
+# Set the default standard to C++11 for all targets.
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+# Do not use, for example, `-std=gnu++11`.
+set(CMAKE_CXX_EXTENSIONS OFF)
+
+if (CMAKE_CXX_COMPILER_ID MATCHES MSVC)
+  # TODO(andschwa): Switch to `/W4` and re-enable possible-loss-of-data warnings.
```
[the remark this answers is not in the document]
Comment (akornatskyy) [reply CMakeLists.txt]: done

File: CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+    /W3
+    # Disable permissiveness.
+    /permissive-
+    # C4244 is a possible loss of data warning for integer conversions.
+    /w44244
+    # C4267 is a possible loss of data warning when converting from `size_t`.
+    /w44267)
+endif ()
+
+# WINDOWS CONFIGURATION.
```
[the remark this answers is not in the document]
Comment (akornatskyy) [reply CMakeLists.txt]: done

File: HOWTO.md [hunk 15]
```diff
@@ -0,0 +1,10 @@
+# CMAKE
```
[the remark this answers is not in the document]
Comment (akornatskyy) [reply HOWTO.md]: removed.

File: CMakeLists.txt (hunk 6 above)
Comment (akornatskyy) [reply CMakeLists.txt]: That is not kind of TODO due to some limitation stated. This cannot be fixed immediately, without touching a larger refactoring field.

File: metrics/isolator.hpp (hunk 8 above)
Comment (akornatskyy) [reply metrics/isolator.hpp]: The reason of `extern "C"` is to have the specific name visible and as such have an ability to load module.

File: overlay/CMakeLists.txt (hunk 9 above)
Comment (akornatskyy) [reply overlay/CMakeLists.txt]: No. That will break linux build.

File: overlay/CMakeLists.txt (hunk 10 above)
Comment (akornatskyy) [reply overlay/CMakeLists.txt]: No. The above is clear to me. If you have a better way to make this more clear, give me some text and I will do replacement to avoid ping-pong.

File: cmake/Dependencies.cmake (hunk 11 above)
Comment (akornatskyy) [reply cmake/Dependencies.cmake]: Yes.

libjournaldlogger.dll 5.0M
logsink.dll 267K
metrics.dll 5.3M

File: CMakeLists.txt (hunk 7 above)
Comment (akornatskyy) [reply CMakeLists.txt]: https://github.com/dcos/dcos-mesos-modules/blob/master/Makefile.am#L192-L198

File: overlay/agent.hpp [hunk 16]
```diff
@@ -13,6 +13,13 @@
 #include <overlay/agent_metrics.hpp>
 #include <overlay/messages.hpp>
 
+// `REGISTERING` is used as an enum value, but it's actually defined as a
+// constant in the Windows SDK.
+#ifdef __WINDOWS__
+#undef REGISTERING
+#undef REGISTERED
+#endif // __WINDOWS__
+
```
[the remark this answers is not in the document]
Comment (asekretenko) [reply overlay/agent.hpp]: Do I understand correctly that the conflicting definitions just come from this file (ManagerProcess::State below), and are not #include`d from some Mesos headers? 

If so, can you rename them now or, at least,  leave `//TODO(akornatskyy): Rename enum values in ManagerProcess::State in this file to avoid name collistion.` ?

File: overlay/agent.cpp [hunk 17]
```diff
@@ -558,7 +559,7 @@ void ManagerProcess::doReliableRegistration(Duration maxBackoff)
 
   // Determine the delay for next attempt by picking a random
   // duration between 0 and 'maxBackoff'.
-  Duration backoff = maxBackoff * ((double) ::random() / RAND_MAX);
+  Duration backoff = maxBackoff * ((double) rand() / RAND_MAX);
```
[the remark this answers is not in the document]
Comment (asekretenko) [reply overlay/agent.cpp]: Could you please either convert these to `std::uniform_real_distribution<double>` now or leave in the code a `//TODO` comment to do this in the future?

File: overlay/agent.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (akornatskyy) [reply overlay/agent.hpp]: added TODO

File: overlay/agent.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (akornatskyy) [reply overlay/agent.cpp]: converted

[the remark this answers is not in the document]
Comment (akornatskyy) [reply overlay/agent.cpp]: I will revert my change since it appears to be compiling on windows but breaks build for linux.

File: overlay/CMakeLists.txt (hunk 10 above)
Comment (asekretenko) [reply overlay/CMakeLists.txt]: Found https://issues.apache.org/jira/browse/MESOS-5932 for porting replicated log to Windows (need to update leveldb first).

I would suggest, instead of putting error message into the comment, to point to the real underlying issue, like
```
# TODO(akornatskyy): Windows build of overlay module master (which depends on Mesos replicated log)
# can be re-enabled after replicated log is available on Windows (see MESOS-5932).
```

File: CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+add_subdirectory(logsink)
+
+# FIXME(akornatskyy): due to import path in 'overlay/messages.proto'
+# NOTE: according to [1], overlay module is not required for Windows
+# [1] https://github.com/mesosphere/dcos-enterprise/tree/master/packages/mesos/extra/windows-ee.patches
+#add_subdirectory(overlay)
+
+
+# FIXME(akornatskyy): due to limitation of protobuf generation process
+# and include in source
```
[the remark this answers is not in the document]
Comment (akornatskyy) [reply CMakeLists.txt]: No, it is not... this has no relation to cmake specifics, the issue is with source code layout and relative paths which incorrectly resolve. I have a task in backlog so that is not an year to get fixed. As part of port to cmake linux, it would require refactor the code a bit to fix.

File: CMakeLists.txt (hunk 7 above)
Comment (akornatskyy) [reply CMakeLists.txt]: done

File: metrics/isolator.hpp (hunk 8 above)
Comment (akornatskyy) [reply metrics/isolator.hpp]: right

File: CMakeLists.txt (hunk 2 above)
Comment (akornatskyy) [reply CMakeLists.txt]: done

File: overlay/CMakeLists.txt (hunk 10 above)
Comment (akornatskyy) [reply overlay/CMakeLists.txt]: done
